ddtrace/internal/telemetry/modules.py

```diff
@@ -0,0 +1,92 @@
+import sys
+from types import ModuleType
+from typing import Any
+from typing import Set
+from typing import Tuple
+
+from ..compat import PYTHON_VERSION_INFO
+from ..module import BaseModuleWatchdog
+
+
+NEW_MODULES: Set[str] = set()  # New modules that have been imported since the last check
+ALL_MODULES: Set[str] = set()  # All modules that have been imported
+MODULE_HOOK_INSTALLED = False
+
+# For Python >= 3.8 we can use the sys.audit event import(module, filename, sys.path, sys.meta_path, sys.path_hooks)
+if PYTHON_VERSION_INFO >= (3, 8):
+
+    def audit_hook(event: str, args: Tuple[Any, ...]):
+        if event != "import":
+            return
+
+        global NEW_MODULES, ALL_MODULES
+        NEW_MODULES.add(args[0])
+        ALL_MODULES.add(args[0])
+
+    def get_newly_imported_modules() -> Set[str]:
+        global MODULE_HOOK_INSTALLED, NEW_MODULES, ALL_MODULES
+
+        # Our hook is not installed, so we are not getting notified of new imports,
+        # we need to track the changes manually
+        if not NEW_MODULES and not MODULE_HOOK_INSTALLED:
+            latest_modules = set(sys.modules.keys())
+            NEW_MODULES = latest_modules - ALL_MODULES
+            ALL_MODULES = latest_modules
+
+        new_modules = NEW_MODULES
+        NEW_MODULES = set()
+        return new_modules
+
+    def install_import_hook():
+        global MODULE_HOOK_INSTALLED, NEW_MODULES, ALL_MODULES
+
+        # If we have not called get_newly_imported_modules yet, we can initialize to all imported modules
+        if not NEW_MODULES:
+            NEW_MODULES = set(sys.modules.keys())
+            ALL_MODULES = NEW_MODULES.copy()
+        sys.addaudithook(audit_hook)
```

Comment:
Double checking my understanding here... The 3 approaches we have are:
1. Periodically go through all imported dependencies (what you originally had)
2. Use ModuleWatchDog which triggers the `after_import` event after each import
3. This approach, where we trigger on all sys audit events, and then check if it's an `import` event type before proceeding.

Naively it feels like the ModuleWatchdog approach is still the most "efficient" in the sense that it only triggers the hook after imports specifically, whereas the sys audit hooks will check for not only `import` events, but also things like `open`, `exec`, etc. Is the perf improvement here based on the assumption that those methods probably won't be called very much, so it's not that much additional performance overhead? Or is there something pretty performance heavy on the ModuleWatchdog side that I'm missing here?

Reply:
Oh wait I think I see my answer:
> The big improvement here is still to just resolve distributions from module names instead of trying to resolve file origin -> distribution.

Reply:
So there are a few pieces.

The main performance improvement is to not try to resolve the module origin. We don't need it for this telemetry data. That part of the code is slow, and it isn't what we actually want for telemetry data. Instead we want the "distribution" not the module (e.g. the `packaging` module comes from the `setuptools` distribution).


The other sys.audit change is because, if we don't need to resolve origin, we don't need all the data provided by ModuleWatchdog. We really only need the module name that was imported to then try to resolve a distribution for it. This way we don't need to worry about installing a loader into the `sys.meta_path`. `sys.audit` will be less invasive and more performant in this case when only collecting the imported module name. So I want to prefer `sys.audit` if it is available.


If we start to use `sys.audit` for more than just this, we may want to centralize a single handler... but... future us problem.